Introduce Atmos YAML functions (#810)
cloudposse/atmos: website/docs/core-concepts/stacks/yaml-functions/yaml-functions.mdx

and user-defined [Explicit Tags](https://yaml.org/spec/1.2.2/#tags) (local data types).
Explicit tags are denoted by the exclamation point (__"!"__) symbol.
Atmos detects the tags in the stack manifests and executes the corresponding functions.

:::info
YAML supports three types of data: core, defined, and user-defined.

- **Core types**: Universally supported, including floats, integers, strings, lists, and maps.  
- **Defined types**: Advanced types like binary data, specified in the YAML standard but not always supported.  
- **User-defined types**: Custom extensions for classes, structures, and functions. Atmos leverages user-defined types to implement its custom functions and extend YAML's capabilities.  
:::

## Use-cases

 - The [__`!template`__](/core-concepts/stacks/yaml-functions/template) YAML function is designed to  
   [evaluate and inject outputs containing maps or lists](/core-concepts/stacks/templates/functions/atmos.Component#handling-outputs-containing-maps-or-lists)  
   into the YAML document, whether generated by the [__`atmos.Component`__](/core-concepts/stacks/templates/functions/atmos.Component) template function or any Go template.

>  - The [__`!exec`__](/core-concepts/stacks/yaml-functions/exec) YAML function is used to execute shell scripts and assign

Reviewer: Will it handle JSON output natively? If so, we should suggest that.

Reviewer: Ok, I think we can skip it here.

---
title: Atmos YAML Functions
sidebar_position: 8
sidebar_label: Atmos YAML Functions
id: yaml-functions
---

import File from '@site/src/components/File'
import PillBox from '@site/src/components/PillBox'
import Intro from '@site/src/components/Intro'
import DocCardList from '@theme/DocCardList'

<PillBox>Advanced</PillBox>

<Intro>
    Atmos [YAML](https://en.wikipedia.org/wiki/YAML) Functions are a crucial part of Atmos stack manifests. They allow
    you to manipulate data and perform operations on the data to customize the stack configurations.
</Intro>

Atmos YAML functions are based on [YAML Explicit typing](https://yaml.org/spec/1.2.2/)
and user-defined [Explicit Tags](https://yaml.org/spec/1.2.2/#tags) (local data types).
Explicit tags are denoted by the exclamation point (__"!"__) symbol.
Atmos detects the tags in the stack manifests and executes the corresponding functions.

:::info

> YAML data types can be divided into three categories: core, defined, and user-defined.
> Core are ones expected to exist in any parser (e.g. floats, ints, strings, lists, maps).
> Many more advanced data types, such as binary data, are defined in the YAML specification but not supported
> in all implementations.
> Finally, YAML defines a way to extend the data type definitions locally to accommodate user-defined classes,
> structures, primitives, and functions.

Reviewer: ```suggestion
YAML supports three types of data: core, defined, and user-defined.

- **Core types**: Universally supported, including floats, integers, strings, lists, and maps.  
- **Defined types**: Advanced types like binary data, specified in the YAML standard but not always supported.  
- **User-defined types**: Custom extensions for classes, structures, and functions. Atmos leverages user-defined types to implement its custom functions and extend YAML's capabilities.  
```

:::

## Use-cases

>  - The [__`!template`__](/core-concepts/stacks/yaml-functions/template) YAML function can be used to
>    [handle template outputs containing maps or lists](/core-concepts/stacks/templates/functions/atmos.Component#handling-outputs-containing-maps-or-lists)
>    returned from the [__`atmos.Component`__](/core-concepts/stacks/templates/functions/atmos.Component) template function

Reviewer: ```suggestion
 - The [__`!template`__](/core-concepts/stacks/yaml-functions/template) YAML function is designed to  
   [evaluate and inject outputs containing maps or lists](/core-concepts/stacks/templates/functions/atmos.Component#handling-outputs-containing-maps-or-lists)  
   into the YAML document, whether generated by the [__`atmos.Component`__](/core-concepts/stacks/templates/functions/atmos.Component) template function or any Go template.
```

 - The [__`!exec`__](/core-concepts/stacks/yaml-functions/exec) YAML function is used to execute shell scripts and assign
   the results to the sections in Atmos stack manifests

 - The [__`!terraform.output`__](/core-concepts/stacks/yaml-functions/terraform.output) YAML function is used to
   [read the outputs (remote state) of components directly in Atmos stack manifests](/core-concepts/components/terraform/remote-state)

:::tip

> You can use [Atmos Stack Manifest Templating](/core-concepts/stacks/templates) and Atmos YAML functions in the same
> stack configurations at the same time.
> Atmos processes the templates first, and then executes the YAML functions, allowing you to provide the parameters to
> the YAML functions dynamically.

Reviewer: ```suggestion
You can combine [Atmos Stack Manifest Templating](/core-concepts/stacks/templates) with Atmos YAML functions within the same stack configuration. Atmos processes templates first, followed by YAML functions, enabling you to dynamically provide parameters to the YAML functions.  
```
